### PR TITLE
feat(Graph2): Add vertex tooltip support

### LINKED ISSUE #3646: Graph2 needs support for tooltips

### PR DESCRIPTION
**Pull request by jbrundage**
Fixes #3646

Signed-off-by: Jaman Brundage <jbrundage372@gmail.com>

## Checklist:
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit message includes a "fixes" reference if appropriate.
  - [x] The commit is signed.
- [x] The change has been fully tested:
  - [x] I have viewed all related gallery items
  - [ ] I have viewed all related dermatology items
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised new issues to address them separately

### CODE REVIEW

File: packages/graph/src/graph2/dataGraph.ts [hunk 1]
```diff
@@ -43,8 +43,6 @@ export class DataGraph extends Graph2 {
     @publish("", "string", "Subgraph Label column")
     subgraphLabelColumn: publish<this, string>;
 
-    @publish([], "any", "Vertex Columns", null, { internal: true })
-    vertexColumns: publish<this, string[]>;
```
Comment (jbrundage) [inline packages/graph/src/graph2/dataGraph.ts:47]: Moved this to Graph2

File: packages/graph/src/graph2/graph.ts [hunk 2]
```diff
@@ -968,6 +987,10 @@ export class Graph2 extends SVGZoomWidget {
         }
     }
 
+    tooltipHTML(data) {
+        return data.props[this.tooltipContentColumn()];
+    }
```
Comment (jbrundage) [inline packages/graph/src/graph2/graph.ts:992]: Users can either overwrite this method or use `tooltipContentColumn` (or both if they want to rely on tooltipContentColumn within their custom tooltipHTML method).

File: packages/graph/src/graph2/dataGraph.ts (hunk 1 above)
Comment (GordonSmith) [reply packages/graph/src/graph2/dataGraph.ts:47]: It doesn't belong in Graph2 as there is no concept of columns / data for the vertices,

IOW you should add the tooltip info to either the react Vertex or the IVertex placeholder (as an optional property).

File: packages/graph/src/graph2/graph.ts [hunk 3]
```diff
@@ -1037,6 +1060,9 @@ export interface Graph2 {
     maxScale(_: number): this;
     showVertexLabels(): boolean;
     showVertexLabels(_: boolean): this;
+    tooltipContentColumn(): string;
+    tooltipContentColumn(_: string): this;
+    tooltipContentColumn_exists(): boolean;
```
Comment (GordonSmith) [inline packages/graph/src/graph2/graph.ts:1065]: Move to GraphData

File: packages/graph/src/graph2/graph.ts [hunk 4]
```diff
@@ -1069,11 +1095,21 @@ export interface Graph2 {
...
-
+    tooltipWidth(): number;
+    tooltipWidth(_: number): this;
+    tooltipHeight(): number;
+    tooltipHeight(_: number): this;
+    enableTooltipPointerEvents(): boolean;
+    enableTooltipPointerEvents(_: boolean): this;
+    vertexColumns(): string[];
+    vertexColumns(_: string[]): this;
+
```
Comment (GordonSmith) [inline packages/graph/src/graph2/graph.ts:1106]: remove VertexColumns

File: packages/graph/src/graph2/graph.ts [hunk 5]
```diff
@@ -1069,11 +1095,21 @@ export interface Graph2 {
...
+    enableTooltipPointerEvents(): boolean;
+    enableTooltipPointerEvents(_: boolean): this;
+    vertexColumns(): string[];
+    vertexColumns(_: string[]): this;
+    
     wasmFolder(): string;
     wasmFolder(_: string): this;
 }
 
+Graph2.prototype.publish("vertexColumns", [], "array", "Vertex Columns", null, { internal: true });
```
Comment (GordonSmith) [inline packages/graph/src/graph2/graph.ts:1111]: Remove vertexColumns (back to GraphData)

File: packages/graph/src/graph2/graph.ts [hunk 6]
```diff
@@ -1069,11 +1095,21 @@ export interface Graph2 {
...
+Graph2.prototype.publish("tooltipContentColumn", null, "string", "Column with tooltip html content", function (this: Graph2) { return this.vertexColumns(); }, { optional: true });
```
Comment (GordonSmith) [inline packages/graph/src/graph2/graph.ts:1112]: Won't be needed when tooltip is added to Vertex or IVertex placeholder

File: packages/html/src/HTMLTooltip.ts [hunk 7]
```diff
@@ -244,6 +246,22 @@ export class HTMLTooltip extends HTMLWidget {
...
+        });
+        this._tooltipElement.on("mouseout", () => {
+            this.mouseout();
+        });
+        setTimeout(()=>{
+            if(this._closing){
+                this.visible(false);
+            }
+        }, this.closeDelay());
+    }
```
Comment (GordonSmith) [inline packages/html/src/HTMLTooltip.ts:263]: What is this fixing?

Comment (jbrundage) [reply packages/html/src/HTMLTooltip.ts:263]: When the cursor leaves the vertex to (potentially) interact with the content of the tooltip `this.closeDelay()` is the number of milliseconds it has to re-enter the vertex or enter the tooltip. (before the vertex disappears)

**Pull request merged**